<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Martin Petr">

<title>slendr simulations</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="mpieva-slendr-2023_files/libs/clipboard/clipboard.min.js"></script>
<script src="mpieva-slendr-2023_files/libs/quarto-html/quarto.js"></script>
<script src="mpieva-slendr-2023_files/libs/quarto-html/popper.min.js"></script>
<script src="mpieva-slendr-2023_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="mpieva-slendr-2023_files/libs/quarto-html/anchor.min.js"></script>
<link href="mpieva-slendr-2023_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="mpieva-slendr-2023_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="mpieva-slendr-2023_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="mpieva-slendr-2023_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="mpieva-slendr-2023_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><em>slendr</em> simulations</h1>
<p class="subtitle lead">practical workshop at MPI EVA Leipzig</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Martin Petr </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">February 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="section" class="level1">
<h1></h1>
<div class="columns">
<div class="column" style="width:70%;">
<blockquote class="blockquote">
<p>Many problems in population genetics cannot be solved by a mathematician, no matter how gifted. [It] is already clear that computer methods are very powerful. This is good. It […] <strong>permits people with limited mathematical knowledge to work on important problems</strong> […]</p>
</blockquote>
</div><div class="column" style="width:30%;">
<p><img src="images/crow.jpeg" class="img-fluid"></p>
<p><a href="https://en.wikipedia.org/wiki/James_F._Crow">James F. Crow</a> – <a href="http://www.gnxp.com/blog/2006/06/10-questions-for-jim-crow.php">interview</a></p>
</div>
</div>
</section>
<section id="why-use-simulations" class="level1 page-columns page-full">
<h1>Why use simulations?</h1>
<ol type="1">
<li>Making sense of estimated statistics</li>
<li>Fitting model parameters</li>
<li>Ground truth for method work</li>
</ol>
<section id="making-sense-of-estimated-statistics" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="making-sense-of-estimated-statistics">Making sense of estimated statistics</h2>
<center>
<img src="images/bacho_kiro.png" class="img-fluid" style="width:70.0%">
</center>

<div class="no-row-height column-margin column-container"><div class="">
<p><a href="https://www.nature.com/articles/s41586-021-03335-3/figures/2">Hajdinjak (2021)</a></p>
</div></div></section>
<section id="making-sense-of-estimated-statistics-1" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="making-sense-of-estimated-statistics-1">Making sense of estimated statistics</h2>
<center>
<img src="images/elephants.png" class="img-fluid" style="width:65.0%">
</center>

<div class="no-row-height column-margin column-container"><div class="">
<p><a href="https://www.pnas.org/doi/10.1073/pnas.1720554115">Palkopoulou (2018)</a></p>
</div></div></section>
<section id="making-sense-of-estimated-statistics-2" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="making-sense-of-estimated-statistics-2">Making sense of estimated statistics</h2>
<center>
<img src="images/neand_decline.png" class="img-fluid" style="width:65.0%">
</center>

<div class="no-row-height column-margin column-container"><div class="">
<p><a href="https://www.pnas.org/doi/abs/10.1073/pnas.1814338116">Petr (2019)</a></p>
</div></div></section>
<section id="fitting-model-parameters-i.e.-abc" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="fitting-model-parameters-i.e.-abc">Fitting model parameters (i.e.&nbsp;<a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">ABC</a>)</h2>
<center>
<img src="images/abc_scheme.png" class="img-fluid" style="width:50.0%">
</center>

<div class="no-row-height column-margin column-container"><div class="">
<p>Image from <a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">Wikipedia on ABC</a></p>
</div></div></section>
<section id="ground-truth-for-methods-work" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="ground-truth-for-methods-work">Ground truth for methods work</h2>
<center>
<img src="images/mcmc.png" class="img-fluid">
</center>

<div class="no-row-height column-margin column-container"><div class="">
<p><a href="https://www.nature.com/articles/ng.3015">Schiffels and Durbin (2014)</a></p>
</div></div></section>
</section>
<section id="what-does-it-mean-to-simulate-a-genome" class="level1">
<h1>What does it mean to simulate a genome?</h1>
<p><br></p>
<div class="incremental">
<p>How would you design an algorithm for a popgen simulation?</p>
<p>Which minimal pieces are needed for the program to be useful?</p>
</div>
</section>
<section id="if-we-want-to-simulate-population-genetics" class="level1">
<h1>If we want to simulate population genetics…</h1>
<p><br></p>
<div class="incremental">
<p>We need <em>populations</em>.</p>
</div>
<div class="incremental">
<p>We need <em>genetics</em>.</p>
<section id="a-chromosome-is" class="level2">
<h2 class="anchored" data-anchor-id="a-chromosome-is">A chromosome is…</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/dna_sequence.png" class="img-fluid figure-img" style="width:50.0%"></p>
</figure>
</div>
<p><br>…a linear sequence of nucleotides…</p>
<div class="incremental">
<ul>
<li>a list of characters (A/G/C/T nucleotides)</li>
<li>a list of 0 or 1 values (ancestral/derived allele)</li>
</ul>
</div>
<div class="incremental">
<p><br></p>
<center>
<strong><em>This could be a simple vector of letters/numbers in R.</em></strong>
</center>
</div>
</section>
<section id="a-population-is" class="level2">
<h2 class="anchored" data-anchor-id="a-population-is">A population is…</h2>
<div class="incremental">
<p><br></p>
<p>A collection of <em>individuals</em> at a given point in time.</p>
</div>
<div class="incremental">
<p>Each carrying a pair of <em>chromosomes</em> inherited from parents.</p>
</div>
<div class="incremental">
<p><br></p>
<center>
<strong><em>This could be a a list of “chromosome vectors” in R…</em></strong>
</center>
</div>
<div class="incremental">
<p><br></p>
<center>
<strong><em>… evolving (in a for loop), from generation to generation.</em></strong>
</center>
</div>
</section>
</div>
</section>
<section id="section-1" class="level1" data-background-image="images/montypython.jpeg">
<h1 data-background-image="images/montypython.jpeg"></h1>
<div>
<h1 color="black" style="background-color: white">
<p>OK. But let’s talk about “real” simulations.</p>
</h1>
</div>
<section id="there-are-many-simulation-tools" class="level2">
<h2 class="anchored" data-anchor-id="there-are-many-simulation-tools">There are many simulation tools</h2>
<p><br></p>
<p>The most famous and widely used are <a href="https://messerlab.org/slim/">SLiM</a> and <a href="https://tskit.dev/msprime/docs/stable/intro.html"><em>msprime</em></a>.</p>
<div class="fragment">
<p><br></p>
<p>Both are very powerful…</p>
<p>… but they require a lot of programming knowledge...</p>
<p>… and a lot of code for non-trivial simulations (🐛🪲🐜).</p>
</div>
<div class="fragment">
<p><br></p>
<center>
<strong>This is why we will focus on <a href="http://www.slendr.net"><em>slendr</em></a>,</strong><br>a convenient R interface to both SLiM and <em>msprime</em>.
</center>
</div>
</section>
</section>
<section id="slim" class="level1">
<h1>SLiM</h1>
<section id="section-2" class="level2">
<h2 class="anchored" data-anchor-id="section-2"></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2 class="anchored">
What is SLiM?
</h2>
<div class="fragment">
<ul>
<li><strong>Forward-time simulator</strong></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>It’s fully programmable!</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Massive library of functions for:
<ul>
<li>Demographic events</li>
<li>Various mating systems</li>
<li>Selection, quantitative traits, …</li>
</ul></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>&gt; 700 pages long <a href="https://github.com/MesserLab/SLiM/releases/download/v3.7.1/SLiM_Manual.pdf">manual</a>!</li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/sim_sketches.001.png" class="img-fluid figure-img" style="width:100.0%"></p>
<p></p><figcaption class="figure-caption">Modified from Alexei Drummond</figcaption><p></p>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="slimgui-ide-for-slim" class="level2">
<h2 class="anchored" data-anchor-id="slimgui-ide-for-slim">SLiMgui – <a href="https://en.wikipedia.org/wiki/Integrated_development_environment">IDE</a> for SLiM</h2>
<center>
<img src="images/slimgui.png" class="img-fluid">
</center>
</section>
<section id="simple-neutral-simulation-in-slim" class="level2">
<h2 class="anchored" data-anchor-id="simple-neutral-simulation-in-slim">Simple neutral simulation in SLiM</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>initialize() {
    // create a neutral mutation type
    initializeMutationType("m1", 0.5, "f", 0.0);
    // initialize 1Mb segment
    initializeGenomicElementType("g1", m1, 1.0);
    initializeGenomicElement(g1, 0, 999999);
    // set mutation rate and recombination rate of the segment
    initializeMutationRate(1e-8);
    initializeRecombinationRate(1e-8);
}

// create an ancestral population p1 of 10000 diploid individuals
1 early() { sim.addSubpop("p1", 10000); }

// in generation 1000, create two daughter populations p2 and p3
1000 early() {
    sim.addSubpopSplit("p2", 5000, p1);
    sim.addSubpopSplit("p3", 1000, p1);
}

// in generation 10000, stop the simulation and save 100 individuals
// from p2 and p3 to a VCF file
10000 late() {
    p2_subset = sample(p2.individuals, 100);
    p3_subset = sample(p3.individuals, 100);
    c(p2_subset, p3_subset).genomes.outputVCF("/tmp/slim_output.vcf.gz");
    sim.simulationFinished();
    catn("DONE!");
}</code></pre>
</div>
</div>
</section>
</section>
<section id="msprime" class="level1 page-columns page-full">
<h1><em>msprime</em></h1>
<section id="section-3" class="level2">
<h2 class="anchored" data-anchor-id="section-3"></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2 class="anchored">
What is <a href="https://tskit.dev/msprime/docs/stable/intro.html"><em>msprime</em></a>?
</h2>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/sim_sketches.001.png" class="img-fluid figure-img" style="width:100.0%"></p>
<p></p><figcaption class="figure-caption">Modified from Alexei Drummond</figcaption><p></p>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="section-4" class="level2">
<h2 class="anchored" data-anchor-id="section-4"></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2 class="anchored">
What is <a href="https://tskit.dev/msprime/docs/stable/intro.html"><em>msprime</em></a>?
</h2>
<div class="fragment">
<ul>
<li>A Python module for writing <strong>coalescent simulations</strong></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Extremely fast (genome-scale, population-scale data)</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>You must know Python fairly well to build complex models</li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/sim_sketches.002.png" class="img-fluid figure-img" style="width:100.0%"></p>
<p></p><figcaption class="figure-caption">Modified from Alexei Drummond</figcaption><p></p>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="simple-simulation-using-msprime" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="simple-simulation-using-msprime">Simple simulation using <em>msprime</em></h2>
<p>This is basically the same model as the SLiM script earlier:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>import msprime

demography = msprime.Demography()
demography.add_population(name="A", initial_size=10_000)
demography.add_population(name="B", initial_size=5_000)
demography.add_population(name="C", initial_size=1_000)
demography.add_population_split(time=1000, derived=["A", "B"], ancestral="C")

ts = msprime.sim_ancestry(
  sequence_length=10e6,
  recombination_rate=1e-8,
  samples={"A": 100, "B": 100},
  demography=demography
)</code></pre>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="">
<p>source: <a href="https://tskit.dev/msprime/docs/stable/demography.html#demographic-models">link</a></p>
</div></div></section>
</section>
<section id="section-5" class="level1">
<h1></h1>
<center>
<p><img src="images/slendr_logo.png" class="img-fluid" style="width:30.0%"></p>
<p><br></p>
<h2 class="anchored" data-anchor-id="section-5">
<a href="https://www.slendr.net">www.slendr.net</a>
</h2>
</center>
<section id="why-a-new-package-spatial-simulations" class="level2">
<h2 class="anchored" data-anchor-id="why-a-new-package-spatial-simulations">Why a new package? – spatial simulations!</h2>
<div class="fragment">
<center>
<img src="images/animation.gif" class="img-fluid" style="width:70.0%">
</center>
</div>
</section>
<section id="why-a-new-package" class="level2">
<h2 class="anchored" data-anchor-id="why-a-new-package">Why a new package?</h2>
<div class="incremental">
<ul>
<li><p>Most researchers are not expert programmers</p></li>
<li><p>All but the most trivial simulations require lots of code</p></li>
</ul>
</div>
<div class="incremental">
<ul>
<li><p>90% <citation needed=""> of simulations are basically the same!</citation></p>
<ul>
<li><p>create populations (splits and <span class="math inline">\(N_e\)</span> changes)</p></li>
<li><p>specify if/how they should mix (rates and times)</p></li>
<li><p>save output (VCF, EIGENSTRAT)</p></li>
</ul></li>
</ul>
</div>
<div class="incremental">
<ul>
<li>Lot of code duplication across projects</li>
</ul>
</div>
</section>
</section>
<section id="lets-get-started" class="level1">
<h1>Let’s get started</h1>
<section id="we-will-need-slendr-tidyverse" class="level2">
<h2 class="anchored" data-anchor-id="we-will-need-slendr-tidyverse">We will need <em>slendr</em> &amp; <em>tidyverse</em></h2>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load data analysis and plotting packages</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)     </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># load slendr itself</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>=======================================================================
NOTE: Due to Python setup issues on some systems which have been
causing trouble particularly for novice users, calling library(slendr)
no longer activates slendr's Python environment automatically.

In order to use slendr's msprime back end or its tree-sequence
functionality, users must now activate slendr's Python environment
manually by executing init_env() after calling library(slendr).

(This note will be removed in the next major version of slendr.)
=======================================================================</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'slendr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:magrittr':

    subtract</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>The interface to all required Python modules has been activated.</code></pre>
</div>
</div>
<center>
<br><em>(ignore the message about missing SLiM)</em>
</center>
</section>
</section>
<section id="section-6" class="level1 page-columns page-full">
<h1></h1>
<div style="text-align: right">
<h1>
<em>slendr</em> haiku
</h1>
</div>
<p><br></p>
<div style="text-align: right">
<p>Build simple models,</p>
<p><br></p>
<p>simulate data from them.</p>
<p><br></p>
<p>Just one plain R script.</p>
</div>
<section id="typical-steps-outline-of-this-tutorial" class="level2">
<h2 class="anchored" data-anchor-id="typical-steps-outline-of-this-tutorial">Typical steps (outline of this tutorial)</h2>
<p><br></p>
<ol type="1">
<li>creating populations</li>
<li>scheduling population splits</li>
<li>programming <span class="math inline">\(N_e\)</span> size changes</li>
<li>encoding gene-flow events</li>
<li>simulation sequence of a given size</li>
<li>computing statistics from simulated outputs</li>
</ol>
</section>
<section id="creating-a-population" class="level2">
<h2 class="anchored" data-anchor-id="creating-a-population">Creating a <code>population()</code></h2>
<p>Each needs a name, size and time of appearance (i.e., “split”):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p><br></p>
<p>This creates a normal R object. Typing it gives a summary:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>pop1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop1 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 1: created as an ancestral population (N = 1000)</code></pre>
</div>
</div>
</div>
</section>
<section id="programming-population-splits" class="level2">
<h2 class="anchored" data-anchor-id="programming-population-splits">Programming population splits</h2>
<p>Splits are indicated by the <code>parent = &lt;pop&gt;</code> argument:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p><br></p>
<p>The split is reported in the “historical summary”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>pop2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop2 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 50: split from pop1 (N = 100)</code></pre>
</div>
</div>
</div>
</section>
<section id="scheduling-resize-events-resize" class="level2">
<h2 class="anchored" data-anchor-id="scheduling-resize-events-resize">Scheduling resize events – <code>resize()</code></h2>
<p>Step size decrease:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb19-2"><a href="#cb19-2"></a>pop1_step <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop1, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p>Exponential increase:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span>
<span id="cb20-2"><a href="#cb20-2"></a>pop2_exp <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop2, <span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="tidyverse-style-pipe-interface" class="level2">
<h2 class="anchored" data-anchor-id="tidyverse-style-pipe-interface">Tidyverse-style <a href="https://magrittr.tidyverse.org">pipe</a> interface</h2>
<p>A more concise way to express the same thing as before.</p>
<p>Step size decrease:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Exponential increase:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="a-more-complex-model" class="level2">
<h2 class="anchored" data-anchor-id="a-more-complex-model">A more complex model</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">300</span>, <span class="at">parent =</span> pop1) <span class="sc">%&gt;%</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">1000</span>)</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>pop3 <span class="ot">&lt;-</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop3"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">400</span>, <span class="at">parent =</span> pop2) <span class="sc">%&gt;%</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">2500</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">800</span>)</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>pop4 <span class="ot">&lt;-</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop4"</span>, <span class="at">N =</span> <span class="dv">1500</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">parent =</span> pop3) <span class="sc">%&gt;%</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">700</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1200</span>, <span class="at">end =</span> <span class="dv">2000</span>)</span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>pop5 <span class="ot">&lt;-</span></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop5"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">600</span>, <span class="at">parent =</span> pop4) <span class="sc">%&gt;%</span></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">50</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">900</span>) <span class="sc">%&gt;%</span></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">250</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">1200</span>) <span class="sc">%&gt;%</span></span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1600</span>, <span class="at">end =</span> <span class="dv">2200</span>) <span class="sc">%&gt;%</span></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">400</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">2400</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="remember-that-slendr-objects-carry-their-history" class="level2">
<h2 class="anchored" data-anchor-id="remember-that-slendr-objects-carry-their-history">Remember that <em>slendr</em> objects carry their history</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>pop5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop5 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 600: split from pop4 (N = 100)
  - time 900: resize from 100 to 50 individuals
  - time 1200: resize from 50 to 250 individuals
  - time 1600-2200: exponential resize from 250 to 1000 individuals
  - time 2400: resize from 1000 to 400 individuals</code></pre>
</div>
</div>
</section>
<section id="last-step-before-simulation-compile_model" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="last-step-before-simulation-compile_model">Last step before simulation: <code>compile_model()</code></h2>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop1, pop2, pop3, pop4, pop5),</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">1</span>,</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">simulation_length =</span> <span class="dv">3000</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong>Compilation takes a list of model components, performs internal consistency checks, returns a single model object.</strong>
</center>

<div class="no-row-height column-margin column-container"><div class="">
<p>The model is also compiled to disk which gives a nice additional layer of reproducibility. The exact location can be specified via <code>path =</code> argument to <code>compile_model()</code>.</p>
</div></div></section>
<section id="model-summary" class="level2">
<h2 class="anchored" data-anchor-id="model-summary">Model summary</h2>
<p>Typing the compiled model prints a brief summary:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'model' object 
--------------------- 
populations: pop1, pop2, pop3, pop4, pop5 
geneflow events: [no geneflow]
generation time: 1 
time direction: forward 
total running length: 3000 model time units
model type: non-spatial

configuration files in: /private/var/folders/d_/hblb15pd3b94rg0v35920wd80000gn/T/RtmpjACFJv/filef7b1674f1fdc </code></pre>
</div>
</div>
</section>
<section id="model-visualization" class="level2">
<h2 class="anchored" data-anchor-id="model-visualization">Model visualization</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">300</span>, <span class="at">parent =</span> pop1) <span class="sc">%&gt;%</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">1000</span>)</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>pop3 <span class="ot">&lt;-</span></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop3"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">400</span>, <span class="at">parent =</span> pop2) <span class="sc">%&gt;%</span></span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">2500</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">800</span>)</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>pop4 <span class="ot">&lt;-</span></span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop4"</span>, <span class="at">N =</span> <span class="dv">1500</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">parent =</span> pop3) <span class="sc">%&gt;%</span></span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">700</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1200</span>, <span class="at">end =</span> <span class="dv">2000</span>)</span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>pop5 <span class="ot">&lt;-</span></span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop5"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">600</span>, <span class="at">parent =</span> pop4) <span class="sc">%&gt;%</span></span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">50</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">900</span>) <span class="sc">%&gt;%</span></span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">250</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">1200</span>) <span class="sc">%&gt;%</span></span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1600</span>, <span class="at">end =</span> <span class="dv">2200</span>) <span class="sc">%&gt;%</span></span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">400</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">2400</span>)</span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb29-23"><a href="#cb29-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop1, pop2, pop3, pop4, pop5),</span>
<span id="cb29-24"><a href="#cb29-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">1</span>,</span>
<span id="cb29-25"><a href="#cb29-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">simulation_length =</span> <span class="dv">3000</span></span>
<span id="cb29-26"><a href="#cb29-26" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="model-visualization-1" class="level2">
<h2 class="anchored" data-anchor-id="model-visualization-1">Model visualization</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mpieva-slendr-2023_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="exercise-1" class="level1 page-columns page-full">
<h1>Exercise #1</h1>
<section id="exercise-1-write-your-own-model" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="exercise-1-write-your-own-model">Exercise #1 — write your own model!</h2>
<div class="columns">
<div class="column" style="width:55%;">
<p>You can use this “template”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>chimp <span class="ot">&lt;-</span> <span class="fu">population</span>(...)</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="co"># &lt;... rest of your code ...&gt;</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">populations =</span> <span class="fu">list</span>(chimp, ...),</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">30</span></span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model) <span class="co"># verify visually</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column" style="width:45%;">
<p><img src="images/intro_model1.png" class="img-fluid"></p>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="">
<p>Don’t worry about gene flow just yet. We will add that at a later stage.</p>
<p>Feel free to include expansions and contractions (maybe in EUR at some point?).</p>
</div></div></section>
</section>
<section id="exercise-1-solution" class="level1">
<h1>Exercise #1 — solution</h1>
<section id="simulating-data-finally" class="level2">
<h2 class="anchored" data-anchor-id="simulating-data-finally">Simulating data (finally…)</h2>
<p>We have a compiled <code>model</code>, how do we simulate data?</p>
<div class="incremental">
<p><em>slendr</em> has two built-in simulation engines:</p>
<ul>
<li>SLiM engine</li>
<li><em>msprime</em> engine</li>
</ul>
</div>
<div class="incremental">
<center>
<h4 class="anchored">
<strong>You don’t have to write any <em>msprime</em> or SLiM code!</strong>
</h4>
</center>
</div>
<div class="incremental">
<p><br> This is all that’s needed:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<center>
(<code>ts</code> is a so-called <strong>tree sequence</strong>)
</center>
</div>
</section>
</section>
<section id="the-output-of-a-slendr-simulation-is-a-tree-sequence" class="level1">
<h1>The output of a <em>slendr</em> simulation is a <strong>tree sequence</strong></h1>
<section id="what-is-tree-sequence" class="level2">
<h2 class="anchored" data-anchor-id="what-is-tree-sequence">What is tree sequence?</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/tree_sequence_diagram.png" class="img-fluid figure-img" style="width:80.0%"></p>
</figure>
</div>
<ul>
<li>a record of full genetic ancestry of a set of samples</li>
<li>an encoding of DNA sequence carried by those samples</li>
<li>an efficient analysis framework</li>
</ul>
</section>
</section>
<section id="why-tree-sequence" class="level1">
<h1>Why tree sequence?</h1>
<p><br></p>
<h3 class="anchored">
Why not VCF, EIGENSTRAT, or a genotype table?
</h3>
<section id="what-we-usually-have" class="level2">
<h2 class="anchored" data-anchor-id="what-we-usually-have">What we usually have</h2>
<center>
<img src="images/vcf_screenshot.png" class="img-fluid" style="width:90.0%">
</center>
</section>
<section id="what-we-usually-want" class="level2">
<h2 class="anchored" data-anchor-id="what-we-usually-want">What we usually <em>want</em></h2>
<p>(As full as possible) a representation of our samples’ history:</p>
<center>
<img src="images/tree_sequence_diagram.png" class="img-fluid">
</center>
<div class="fragment">
<center>
<strong>This is exactly what a tree sequence <em>is</em>.</strong>
</center>
</div>
</section>
<section id="tree-sequences-are-also-very-efficient" class="level2">
<h2 class="anchored" data-anchor-id="tree-sequences-are-also-very-efficient">Tree sequences are also very efficient</h2>
<p><br></p>
<p>Let’s take this minimalistic model:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>pop <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop"</span>, <span class="at">time =</span> <span class="fl">1e6</span>, <span class="at">N =</span> <span class="dv">10000</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(pop, <span class="at">generation_time =</span> <span class="dv">30</span>, <span class="at">direction =</span> <span class="st">"backward"</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">1e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>(simulates 2 <span class="math inline">\(\times\)</span> 10000 chromosomes of 100 Mb)</p>
<div class="incremental">
<p><br></p>
<p><strong>Runs in less than 30 seconds on my laptop!</strong></p>
<p><strong>Taking about 66 Mb of memory!</strong></p>
</div>
</section>
<section id="how-does-this-work" class="level2">
<h2 class="anchored" data-anchor-id="how-does-this-work">How does this work?!</h2>
<div class="incremental">
<center>
<p><img src="images/tables.jpeg" class="img-fluid"></p>
<center>
</center></center></div>
</section>
<section id="tree-sequence-tables-tskit-docs" class="level2">
<h2 class="anchored" data-anchor-id="tree-sequence-tables-tskit-docs">Tree-sequence tables (<a href="https://tskit.dev/tutorials/tables_and_editing.html">tskit docs</a>)</h2>
<div class="row">
<div class="columns">
<div class="column" style="width:60%;">
<p>A tree (sequence) can be represented by</p>
<div class="incremental">
<ul class="incremental">
<li>a table of <font color="orange">n</font><font color="green">o</font><font color="darkblue">d</font><font color="green">e</font><font color="darkblue">s</font>,</li>
<li>a table of <u>edges</u> between nodes,</li>
<li>a table of <font color="red">mutations</font> on edges</li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<p><br></p>
<center>
<img src="images/tree_diagram.png" class="img-fluid">
</center>
</div>
</div>
</div>
<div class="incremental">
<center>
<h2 class="anchored">
<strong>A set of such tables is a tree sequence.</strong>
</h2>
</center>
</div>
</section>
<section id="tree-sequence-tables-in-practice" class="level2">
<h2 class="anchored" data-anchor-id="tree-sequence-tables-in-practice">Tree-sequence tables in practice</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>ggtree v3.6.2 For help: https://yulab-smu.top/treedata-book/

If you use the ggtree package suite in published research, please cite
the appropriate paper(s):

Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam.
ggtree: an R package for visualization and annotation of phylogenetic
trees with their covariates and other associated data. Methods in
Ecology and Evolution. 2017, 8(1):28-36. doi:10.1111/2041-210X.12628

Shuangbin Xu, Lin Li, Xiao Luo, Meijun Chen, Wenli Tang, Li Zhan, Zehan
Dai, Tommy T. Lam, Yi Guan, Guangchuang Yu. Ggtree: A serialized data
object for visualization of a phylogenetic tree and annotation data.
iMeta 2022, 4(1):e56. doi:10.1002/imt2.56

Guangchuang Yu. Using ggtree to visualize data on tree-like structures.
Current Protocols in Bioinformatics. 2020, 69:e96. doi:10.1002/cpbi.96</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'ggtree'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:magrittr':

    inset</code></pre>
</div>
<div class="cell-output-display">
<p><img src="mpieva-slendr-2023_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div><div class="column" style="width:50%;">
<p>. . .</p>
<p>nodes:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>  node_id pop_id       time
1      88      0 -348.38998
2      85      0  -19.46493
3      73      0  235.58011</code></pre>
</div>
</div>
<p>. . .</p>
<p>edges:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>  child_node_id parent_node_id
1            85             88
2            73             85
3            53             88</code></pre>
</div>
</div>
<p>. . .</p>
<p>mutations:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] id   site node time
&lt;0 rows&gt; (or 0-length row.names)</code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="lets-take-the-model-we-defined-earlier" class="level2">
<h2 class="anchored" data-anchor-id="lets-take-the-model-we-defined-earlier">Let’s take the <code>model</code> we defined earlier…</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mpieva-slendr-2023_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="and-simulate-tree-sequence-from-it" class="level2">
<h2 class="anchored" data-anchor-id="and-simulate-tree-sequence-from-it">… and simulate tree sequence from it</h2>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">1e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
(A tree sequence can be saved to disk with <code>ts_save()</code>.)
</center>
<div class="incremental">
<p><br></p>
<p>If we type <code>ts</code> into an R console, we get…</p>
</div>
</section>
<section id="a-tree-sequence-content-summary" class="level2">
<h2 class="anchored" data-anchor-id="a-tree-sequence-content-summary">A tree-sequence content summary</h2>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>ts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │       1450║
╟───────────────┼───────────╢
║Sequence Length│    1000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │       9400║
╟───────────────┼───────────╢
║Total Size     │    1.6 MiB║
╚═══════════════╧═══════════╝
╔═══════════╤═════╤═════════╤════════════╗
║Table      │Rows │Size     │Has Metadata║
╠═══════════╪═════╪═════════╪════════════╣
║Edges      │24445│763.9 KiB│          No║
╟───────────┼─────┼─────────┼────────────╢
║Individuals│ 4700│128.5 KiB│          No║
╟───────────┼─────┼─────────┼────────────╢
║Migrations │    0│  8 Bytes│          No║
╟───────────┼─────┼─────────┼────────────╢
║Mutations  │    0│ 16 Bytes│          No║
╟───────────┼─────┼─────────┼────────────╢
║Nodes      │19989│546.6 KiB│          No║
╟───────────┼─────┼─────────┼────────────╢
║Populations│    5│383 Bytes│         Yes║
╟───────────┼─────┼─────────┼────────────╢
║Provenances│    1│  3.8 KiB│          No║
╟───────────┼─────┼─────────┼────────────╢
║Sites      │    0│ 16 Bytes│          No║
╚═══════════╧═════╧═════════╧════════════╝</code></pre>
</div>
</div>
<!-- ## Saving a tree-sequence permanently -->
<!-- <br> -->
<!-- We can either specify an output file: -->
<!-- ```{r} -->
<!-- #| eval: false -->
<!-- ts <- msprime(model, sequence_length = 1e6, recombination_rate = 1e-8, -->
<!--               output = "<path>") -->
<!-- ``` -->
<!-- <br> -->
<!-- Alternatively, we can always save an already created tree-sequence object: -->
<!-- ```{r} -->
<!-- #| eval: false -->
<!-- ts_save(ts, file = "<path>") -->
<!-- ``` -->
</section>
</section>
<section id="what-can-we-do-with-it" class="level1 page-columns page-full">
<h1>What can we do with it?</h1>
<section id="slendrs-r-interface-to-tskit" class="level2">
<h2 class="anchored" data-anchor-id="slendrs-r-interface-to-tskit"><em>slendr</em>’s R interface to <a href="https://tskit.dev/tskit"><em>tskit</em></a></h2>
<center>
<p><img src="images/slendr_tskit.png" class="img-fluid"></p>
<p>This <a href="https://www.slendr.net/reference/index.html#tree-sequence-loading-and-processing">R interface</a> links to Python methods implemented in <a href="https://tskit.dev/tskit/docs/stable/python-api.html#statistics"><em>tskit</em></a>.</p>
</center>
</section>
<section id="here-is-the-magic" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="here-is-the-magic">Here is the magic</h2>
<p>Tree sequences make it possible to directly compute many quantities of interest <em>without going via conversion to a genotype table/VCF</em>!</p>
<center>
<img src="images/tree_sequence_diagram.png" class="img-fluid">
</center>

<div class="no-row-height column-margin column-container"><div class="">
<p><a href="https://academic.oup.com/genetics/article/215/3/779/5930459">Ralph et al.&nbsp;(2020)</a></p>
</div></div></section>
<section id="in-fact-we-dont-even-need-mutations" class="level2">
<h2 class="anchored" data-anchor-id="in-fact-we-dont-even-need-mutations">In fact, we don’t even <em>need</em> mutations!</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>ts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │       1450║
╟───────────────┼───────────╢
║Sequence Length│    1000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │       9400║
╟───────────────┼───────────╢
║Total Size     │    1.6 MiB║
╚═══════════════╧═══════════╝
╔═══════════╤═════╤═════════╤════════════╗
║Table      │Rows │Size     │Has Metadata║
╠═══════════╪═════╪═════════╪════════════╣
║Edges      │24445│763.9 KiB│          No║
╟───────────┼─────┼─────────┼────────────╢
║Individuals│ 4700│128.5 KiB│          No║
╟───────────┼─────┼─────────┼────────────╢
║Migrations │    0│  8 Bytes│          No║
╟───────────┼─────┼─────────┼────────────╢
║Mutations  │    0│ 16 Bytes│          No║
╟───────────┼─────┼─────────┼────────────╢
║Nodes      │19989│546.6 KiB│          No║
╟───────────┼─────┼─────────┼────────────╢
║Populations│    5│383 Bytes│         Yes║
╟───────────┼─────┼─────────┼────────────╢
║Provenances│    1│  3.8 KiB│          No║
╟───────────┼─────┼─────────┼────────────╢
║Sites      │    0│ 16 Bytes│          No║
╚═══════════╧═════╧═════════╧════════════╝</code></pre>
</div>
</div>
</section>
<section id="how-can-we-compute-statistics" class="level2">
<h2 class="anchored" data-anchor-id="how-can-we-compute-statistics">How can we compute statistics?</h2>
<p>There is a duality between mutations and branch lengths in trees (more <a href="https://tskit.dev/tskit/docs/stable/stats.html">here</a>).</p>
<center>
<p><img src="images/tree_sequence_diagram.png" class="img-fluid"></p>
<h3 class="anchored">
But what if we want mutations?
</h3>
</center>
</section>
<section id="coalescent-and-mutation-processes-can-be-decoupled" class="level2">
<h2 class="anchored" data-anchor-id="coalescent-and-mutation-processes-can-be-decoupled">Coalescent and mutation processes can be decoupled!</h2>
<center>
<p><img src="images/tree_sequence_diagram.png" class="img-fluid"></p>
<h3 class="anchored">
This means we can add mutations<br><em>after</em> the simulation.
</h3>
</center>
</section>
<section id="this-allows-efficient-massive-simulations" class="level2">
<h2 class="anchored" data-anchor-id="this-allows-efficient-massive-simulations">This allows efficient, massive simulations</h2>
<p>If we have a simulated <code>ts</code> object, we can do:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>ts_mutated <span class="ot">&lt;-</span> <span class="fu">ts_mutate</span>(ts, <span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Or, with a shortcut:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
We will be using <code>ts_mutate()</code> throughout.
</center>
</section>
</section>
<section id="how-do-we-use-tree-sequences-in-practice" class="level1">
<h1>How do we use tree sequences in practice?</h1>
<div class="incremental">
<p><br></p>
<h3 class="anchored">
First, to analyze data, we need to be able to refer to “samples”.
</h3>
<section id="extracting-sample-information" class="level2">
<h2 class="anchored" data-anchor-id="extracting-sample-information">Extracting sample information</h2>
<p>Each “sampled” individual in <em>slendr</em> has a symbolic name, a sampling time, and a population assignment.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mpieva-slendr-2023_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="extracting-sample-information-1" class="level2">
<h2 class="anchored" data-anchor-id="extracting-sample-information-1">Extracting sample information</h2>
<p>If we have a tree sequence <code>ts</code>, we can get samples with <code>ts_samples()</code>:</p>
<div class="columns">
<div class="column" style="width:55%;">
<div class="cell" data-output-location="fragment">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4,700 × 3
   name     time pop  
   &lt;chr&gt;   &lt;dbl&gt; &lt;chr&gt;
 1 pop1_1      0 pop1 
 2 pop1_2      0 pop1 
 3 pop1_3      0 pop1 
 4 pop1_4      0 pop1 
 5 pop1_5      0 pop1 
 6 pop1_6      0 pop1 
 7 pop1_7      0 pop1 
 8 pop1_8      0 pop1 
 9 pop1_9      0 pop1 
10 pop1_10     0 pop1 
# … with 4,690 more rows</code></pre>
</div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:43%;">
<div class="cell" data-output-location="fragment">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">count</span>(pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 2
  pop       n
  &lt;chr&gt; &lt;int&gt;
1 pop1   1000
2 pop2    100
3 pop3   2500
4 pop4    700
5 pop5    400</code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="analyzing-tree-sequences-with-slendr" class="level2">
<h2 class="anchored" data-anchor-id="analyzing-tree-sequences-with-slendr">Analyzing tree sequences with <em>slendr</em></h2>
<p>Let’s say we have the following model and we simulate a tree sequence from it.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1"></a>pop <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop"</span>, <span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb51-2"><a href="#cb51-2"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(pop, <span class="at">generation_time =</span> <span class="dv">1</span>, <span class="at">simulation_length =</span> <span class="dv">10000</span>)</span>
<span id="cb51-3"><a href="#cb51-3"></a></span>
<span id="cb51-4"><a href="#cb51-4"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong>Imagine we want to get the allele-frequency spectrum.</strong>
</center>
</section>
<section id="example-allele-frequency-spectrum" class="level2">
<h2 class="anchored" data-anchor-id="example-allele-frequency-spectrum">Example: allele frequency spectrum</h2>
<div class="cell">

</div>
<div class="columns">
<div class="column" style="width:55%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1"></a><span class="co"># sample 5 individuals</span></span>
<span id="cb52-2"><a href="#cb52-2"></a><span class="co"># (i.e. 10 chromosomes)</span></span>
<span id="cb52-3"><a href="#cb52-3"></a>samples <span class="ot">&lt;-</span></span>
<span id="cb52-4"><a href="#cb52-4"></a>  <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span></span>
<span id="cb52-5"><a href="#cb52-5"></a>  <span class="fu">sample_n</span>(<span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb52-6"><a href="#cb52-6"></a>  <span class="fu">pull</span>(name)</span>
<span id="cb52-7"><a href="#cb52-7"></a></span>
<span id="cb52-8"><a href="#cb52-8"></a><span class="co"># compute allele frequency</span></span>
<span id="cb52-9"><a href="#cb52-9"></a><span class="co"># spectrum from the given set</span></span>
<span id="cb52-10"><a href="#cb52-10"></a><span class="co"># of individuals</span></span>
<span id="cb52-11"><a href="#cb52-11"></a>afs1 <span class="ot">&lt;-</span> <span class="fu">ts_afs</span>(</span>
<span id="cb52-12"><a href="#cb52-12"></a>  ts, <span class="fu">list</span>(samples),</span>
<span id="cb52-13"><a href="#cb52-13"></a>  <span class="at">mode =</span> <span class="st">"branch"</span>,</span>
<span id="cb52-14"><a href="#cb52-14"></a>  <span class="at">polarised =</span> <span class="cn">TRUE</span>,</span>
<span id="cb52-15"><a href="#cb52-15"></a>  <span class="at">span_normalise =</span> <span class="cn">TRUE</span></span>
<span id="cb52-16"><a href="#cb52-16"></a>)</span>
<span id="cb52-17"><a href="#cb52-17"></a></span>
<span id="cb52-18"><a href="#cb52-18"></a>afs1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 40374.890 19702.927 13400.619  9944.322  8124.764  6557.752  5643.803
 [8]  4919.553  4444.428  4166.190</code></pre>
</div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:43%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs1, <span class="at">type =</span> <span class="st">"b"</span>,</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"allele count bin"</span>,</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="mpieva-slendr-2023_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="example-allele-frequency-spectrum-1" class="level2">
<h2 class="anchored" data-anchor-id="example-allele-frequency-spectrum-1">Example: allele frequency spectrum</h2>
<div class="cell">

</div>
<div class="columns">
<div class="column" style="width:55%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sample 5 individuals</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (i.e. 10 chromosomes)</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(name)</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a><span class="co"># compute allele frequency</span></span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a><span class="co"># spectrum from the given set</span></span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a><span class="co"># of individuals</span></span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a>afs2 <span class="ot">&lt;-</span> <span class="fu">ts_afs</span>(</span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a>  ts, <span class="fu">list</span>(samples),</span>
<span id="cb55-13"><a href="#cb55-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">polarised =</span> <span class="cn">TRUE</span></span>
<span id="cb55-14"><a href="#cb55-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb55-15"><a href="#cb55-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-16"><a href="#cb55-16" aria-hidden="true" tabindex="-1"></a>afs2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 0 0 0 0 0 0 0 0 0 0</code></pre>
</div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:43%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs2, <span class="at">type =</span> <span class="st">"b"</span>,</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"allele count bin"</span>,</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"frequency"</span>)</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(afs1, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"mutation-based AFS"</span>, <span class="st">"theoretical branch-based AFS"</span>),</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="mpieva-slendr-2023_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
</div>
</section>
</div>
</section>
<section id="exercise-2" class="level1">
<h1>Exercise #2</h1>
<section id="exercise-2-estimating-n_e-from-afs" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2-estimating-n_e-from-afs">Exercise #2 — estimating <span class="math inline">\(N_e\)</span> from AFS</h2>
<div class="cell">

</div>
<p>You sequenced 10 individuals from one population and computed this AFS (counts of singletons, doubletons, …):</p>
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- dput(as.vector(observed_afs)) -->
<!-- ``` -->
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>afs_observed <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">2520</span>, <span class="dv">1449</span>, <span class="dv">855</span>, <span class="dv">622</span>, <span class="dv">530</span>, <span class="dv">446</span>, <span class="dv">365</span>, <span class="dv">334</span>, <span class="dv">349</span>, <span class="dv">244</span>,</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>                  <span class="dv">264</span>, <span class="dv">218</span>,  <span class="dv">133</span>, <span class="dv">173</span>, <span class="dv">159</span>, <span class="dv">142</span>, <span class="dv">167</span>, <span class="dv">129</span>, <span class="dv">125</span>, <span class="dv">143</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p>You know from fossil evidence that the population had constant <span class="math inline">\(N_e\)</span> for the past 100,000 generations, and that the <span class="math inline">\(N_e\)</span> was somewhere between 1000 and 30000.</p>
</div>
<div class="incremental">
<p><strong>Use <em>slendr</em> to guess the true value of</strong> <span class="math inline">\(N_e\)</span> given the observed AFS by running single-population simulations of different <span class="math inline">\(N_e\)</span> and comparing <code>ts_afs()</code> results to <code>afs_observed</code>.</p>
</div>
</section>
<section id="exercise-2-hints" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2-hints">Exercise #2 – hints</h2>
<ol type="1">
<li><p>Write an R function (in a new script) that gets <span class="math inline">\(N_e\)</span> as input, creates a <em>slendr</em> population, compiles a model, simulates a tree sequence, runs <code>ts_afs()</code> on it, and returns the AFS.</p></li>
<li><p>Find the <span class="math inline">\(N_e\)</span> value that will give the closest AFS to the observed one. For instance, you could:</p>
<ul>
<li><p>a ) Plot simulated AFS for different <span class="math inline">\(N_e\)</span> with the AFS and just eyeball <span class="math inline">\(N_e\)</span> value that looks correct from a graph.</p></li>
<li><p>b ) Simulate AFS automatically in steps of possible <span class="math inline">\(N_e\)</span> values and find the <a href="https://en.wikipedia.org/wiki/Mean_squared_error">closest matching</a> one.</p></li>
</ul></li>
</ol>
</section>
<section id="exercise-2-solution-a-eye-balling" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2-solution-a-eye-balling">Exercise #2 solution (a) – eye-balling</h2>
</section>
<section id="exercise-2-solution-b-grid" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2-solution-b-grid">Exercise #2 solution (b) – grid</h2>
</section>
</section>
<section id="gene-flow-admixture" class="level1">
<h1>Gene flow / admixture</h1>
<section id="gene_flow-events" class="level2">
<h2 class="anchored" data-anchor-id="gene_flow-events"><code>gene_flow()</code> events</h2>
<p>Gene flow is programmed the <code>gene_flow()</code> function.</p>
<p><br></p>
<p>If we have populations <code>p1</code> and <code>p2</code>, we schedule gene flow with:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">gene_flow</span>(<span class="at">from =</span> p1, <span class="at">to =</span> p2, <span class="at">start =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">600</span>, <span class="at">rate =</span> <span class="fl">0.13</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p>Multiple gene-flow events can be gathered in a list:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> p1, <span class="at">to =</span> p2, <span class="at">start =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">600</span>, <span class="at">rate =</span> <span class="fl">0.13</span>),</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> <span class="sc">&lt;</span>..<span class="sc">&gt;</span>, <span class="at">to =</span> <span class="sc">&lt;</span>..<span class="sc">&gt;</span>, <span class="at">start =</span> <span class="sc">&lt;</span>...<span class="sc">&gt;</span>, <span class="at">end =</span> <span class="sc">&lt;</span>...<span class="sc">&gt;</span>, <span class="at">rate =</span> <span class="sc">&lt;</span>...<span class="sc">&gt;</span>),</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>  <span class="sc">&lt;</span> potentially many more ... <span class="sc">&gt;</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="incremental">
<p><br></p>
<center>
<strong><code>gene_flow()</code> checks for consistency!</strong>
</center>
<!-- ## Behind the scenes -->
<!-- <br> -->
<!-- The output of `gene_flow()` is nothing but a data frame: -->
<!-- ```{r} -->
<!-- gf -->
<!-- ``` -->
<!-- However, the function does lots of consistency checks behind the scenes, so its better to always use it. -->
</div>
</section>
</section>
<section id="exercise-3" class="level1">
<h1>Exercise #3</h1>
<section id="exercise-3-adding-gene_flow" class="level2">
<h2 class="anchored" data-anchor-id="exercise-3-adding-gene_flow">Exercise #3 — adding <code>gene_flow()</code></h2>
<p>Add some gene-flow events into your model from Exercise #1.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mpieva-slendr-2023_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="exercise-3-adding-gene_flow-1" class="level2">
<h2 class="anchored" data-anchor-id="exercise-3-adding-gene_flow-1">Exercise #3 — adding <code>gene_flow()</code></h2>
<p>Add some gene-flow events into your model from Exercise #1.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mpieva-slendr-2023_files/figure-html/unnamed-chunk-46-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="exercise-4" class="level1">
<h1>Exercise 4</h1>
<section id="exercise-4-popgen-statistics" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4-popgen-statistics">Exercise #4 — popgen statistics</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mpieva-slendr-2023_files/figure-html/unnamed-chunk-47-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="exercise-4-popgen-statistics-1" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4-popgen-statistics-1">Exercise #4 — popgen statistics</h2>
<p>Compute <a href="https://en.wikipedia.org/wiki/Nucleotide_diversity">nucleotide diversity (<span class="math inline">\(\pi\)</span>)</a> in each population (<code>ts_diversity()</code>). Do your results with known <span class="math inline">\(N_e\)</span> values?</p>
<hr>
<p>Compute <code>ts_divergence()</code> (genetic divergence) or <code>ts_fst()</code> (<a href="https://en.wikipedia.org/wiki/Fixation_index"><span class="math inline">\(F_{ST}\)</span></a>) between populations. Do the results match the model?</p>
<hr>
<p>Compute the “outgroup <span class="math inline">\(f_3\)</span>” using CHIMP as the outgroup (<code>ts_f3()</code>) for different pairs of “A” and “B” populations.</p>
<hr>
<p>Test Neanderthal introgression in EUR using the <span class="math inline">\(f_4\)</span> statistic (<code>ts_f4()</code>). [hint: <span class="math inline">\(f_4\)</span>(AFR, EUR; NEAND, CHIMP)]</p>
</section>
<section id="exercise-4-popgen-statistics-2" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4-popgen-statistics-2">Exercise #4 — popgen statistics</h2>
<p>A useful R trick for handling recorded samples:</p>
<div class="cell">

</div>
<div class="columns">
<div class="column" style="width:40%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 24,000 × 3
   name    time pop  
   &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 AFR_1      0 AFR  
 2 AFR_2      0 AFR  
 3 AFR_3      0 AFR  
 4 AFR_4      0 AFR  
 5 AFR_5      0 AFR  
 6 AFR_6      0 AFR  
 7 AFR_7      0 AFR  
 8 AFR_8      0 AFR  
 9 AFR_9      0 AFR  
10 AFR_10     0 AFR  
# … with 23,990 more rows</code></pre>
</div>
</div>
</div><div class="column" style="width:60%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">split</span>(., .<span class="sc">$</span>pop) <span class="sc">%&gt;%</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(pull, <span class="st">"name"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 4
 $ AFR  : chr [1:15000] "AFR_1" "AFR_2" "AFR_3" "AFR_4" ...
 $ CHIMP: chr [1:5000] "CHIMP_1" "CHIMP_2" "CHIMP_3" "CHIMP_4" ...
 $ EUR  : chr [1:3000] "EUR_1" "EUR_2" "EUR_3" "EUR_4" ...
 $ NEA  : chr [1:1000] "NEA_1" "NEA_2" "NEA_3" "NEA_4" ...</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="time-series-data" class="level1">
<h1>Time-series data</h1>
<section id="sampling-adna-samples-through-time" class="level2">
<h2 class="anchored" data-anchor-id="sampling-adna-samples-through-time">Sampling aDNA samples through time</h2>
<p><br></p>
<p>Imagine we have <code>pop1</code>, <code>pop2</code>, … compiled in a <code>model</code>.</p>
<p><br></p>
<p>To record <em>ancient</em> individuals in the tree sequence, we can use <code>schedule_sampling()</code> like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">schedule_sampling</span>(</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>  model,                <span class="co"># compiled slendr model object</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">100</span>, <span class="dv">500</span>),  <span class="co"># at these times (can be also a single number) ...</span></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop1, <span class="dv">42</span>),       <span class="co"># ... sample 42 individuals from pop1</span></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop2, <span class="dv">10</span>),       <span class="co"># ... sample 10 individuals from pop2</span></span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop3, <span class="dv">1</span>)         <span class="co"># ... sample 1 individual from pop 3</span></span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="enforcing-timing-correctness" class="level2">
<h2 class="anchored" data-anchor-id="enforcing-timing-correctness">Enforcing timing correctness</h2>
<p><br></p>
<p><code>schedule_sampling()</code> will only perform sampling from a population if it exists at that time.</p>
<p><br></p>
<p>For instance, we can’t sample modern human from 1 Mya:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">schedule_sampling</span>(</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>  model,</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">times =</span> <span class="fl">1e6</span>,   <span class="co"># can't sample EUR</span></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(eur, <span class="dv">1</span>)   <span class="co"># individual at 1 Mya!</span></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><small> <code> Warning message: No valid sampling events were retained </code> </small></p>
</section>
<section id="sampling-schedule-format" class="level2">
<h2 class="anchored" data-anchor-id="sampling-schedule-format">Sampling schedule format</h2>
<p>The output of <code>schedule_sampling()</code> is a plain data frame!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">40000</span>, <span class="dv">30000</span>, <span class="dv">20000</span>, <span class="dv">10000</span>), <span class="fu">list</span>(eur, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 7
   time pop       n y_orig x_orig y     x    
  &lt;int&gt; &lt;chr&gt; &lt;int&gt; &lt;lgl&gt;  &lt;lgl&gt;  &lt;lgl&gt; &lt;lgl&gt;
1 10000 EUR       1 NA     NA     NA    NA   
2 20000 EUR       1 NA     NA     NA    NA   
3 30000 EUR       1 NA     NA     NA    NA   
4 40000 EUR       1 NA     NA     NA    NA   </code></pre>
</div>
</div>
<div class="incremental">
<p>We can bind multiple sampling schedules together, giving us finer control about sampling:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>eur_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">40000</span>, <span class="dv">30000</span>, <span class="dv">20000</span>, <span class="dv">10000</span>, <span class="dv">0</span>), <span class="fu">list</span>(eur, <span class="dv">1</span>))</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>afr_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="dv">0</span>, <span class="fu">list</span>(afr, <span class="dv">1</span>))</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(eur_samples, afr_samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="how-to-use-a-sampling-schedule" class="level2">
<h2 class="anchored" data-anchor-id="how-to-use-a-sampling-schedule">How to use a sampling schedule?</h2>
<p>To sample individuals based on a given schedule, we use the <code>samples =</code> argument of the <code>msprime()</code> functions:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">samples =</span> samples, <span class="at">sequence_length =</span> <span class="fl">1e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p>We can verify that only specific individuals are recorded:</p>
<div class="columns">
<div class="column" style="width:40%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  name   time pop  
  &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;
1 EUR_1 40000 EUR  
2 EUR_2 30000 EUR  
3 EUR_3 20000 EUR  
4 EUR_4 10000 EUR  
5 AFR_1     0 AFR  
6 EUR_5     0 EUR  </code></pre>
</div>
</div>
</div><div class="column" style="width:60%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>ts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │       1331║
╟───────────────┼───────────╢
║Sequence Length│    1000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │         12║
╟───────────────┼───────────╢
║Total Size     │  289.8 KiB║
╚═══════════════╧═══════════╝
╔═══════════╤════╤═════════╤════════════╗
║Table      │Rows│Size     │Has Metadata║
╠═══════════╪════╪═════════╪════════════╣
║Edges      │4085│127.7 KiB│          No║
╟───────────┼────┼─────────┼────────────╢
║Individuals│   6│192 Bytes│          No║
╟───────────┼────┼─────────┼────────────╢
║Migrations │   0│  8 Bytes│          No║
╟───────────┼────┼─────────┼────────────╢
║Mutations  │1649│ 59.6 KiB│          No║
╟───────────┼────┼─────────┼────────────╢
║Nodes      │ 945│ 25.8 KiB│          No║
╟───────────┼────┼─────────┼────────────╢
║Populations│   4│341 Bytes│         Yes║
╟───────────┼────┼─────────┼────────────╢
║Provenances│   2│  3.5 KiB│          No║
╟───────────┼────┼─────────┼────────────╢
║Sites      │1647│ 40.2 KiB│          No║
╚═══════════╧════╧═════════╧════════════╝</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="exercise-5" class="level1 page-columns page-full">
<h1>Exercise #5</h1>
<section id="exercise-5a-ancient-samples" class="level2">
<h2 class="anchored" data-anchor-id="exercise-5a-ancient-samples">Exercise #5a — ancient samples</h2>
<div class="cell">
<div class="cell-output-display">
<p><img src="mpieva-slendr-2023_files/figure-html/unnamed-chunk-59-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="exercise-5a-ancient-samples-1" class="level2">
<h2 class="anchored" data-anchor-id="exercise-5a-ancient-samples-1">Exercise #5a — ancient samples</h2>
<p>Simulate a new data set from your model, but this time specify a sampling schedule:</p>
<ul>
<li>one present-day CHIMP and AFR individual</li>
<li>20 present-day EUR individuals</li>
<li>1 NEA at 70 ky, 1 NEA at 40 ky</li>
<li>1 EUR every 1000 years between 50-5 kya</li>
</ul>
<p>Reminder: you can do this by:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="co"># rbind(...) together individual sampling schedule data frames</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msprime</span>(model, <span class="at">samples =</span> samples, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="exercise-5b-f_4-ratio-statistic" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="exercise-5b-f_4-ratio-statistic">Exercise #5b — <span class="math inline">\(f_4\)</span>-ratio statistic</h2>
<p>Use <span class="math inline">\(f_4\)</span>-ratio statistic to replicate the <a href="https://www.pnas.org/doi/10.1073/pnas.1814338116#fig01">following figure</a>:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/neand_decline.png" class="img-fluid figure-img" style="width:60.0%"></p>
</figure>
</div>

<div class="no-row-height column-margin column-container"><div class="">
<p>Hint: You can compute Neanderthal ancestry for a vector of individuals <code>X</code> as <code>ts_f4ratio(ts, X = X, "NEA_1", "NEA_2", "AFR_1", "CHIMP_1")</code>.</p>
</div></div></section>
</section>
<section id="exercise-5-solution" class="level1">
<h1>Exercise #5 — solution</h1>
</section>
<section id="bonus-content" class="level1">
<h1>BONUS CONTENT</h1>
<section id="conversion-to-other-genotype-formats" class="level2">
<h2 class="anchored" data-anchor-id="conversion-to-other-genotype-formats">Conversion to other genotype formats</h2>
<p>If you have a tree-sequence object <code>ts</code>…</p>
<div class="incremental">
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_vcf</span>(ts, <span class="at">path =</span> <span class="st">"path/to/a/file.vcf.gz"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="incremental">
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_eigenstrat</span>(ts, <span class="at">prefix =</span> <span class="st">"path/to/eigenstrat/prefix"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="incremental">
<div class="cell">

</div>
<div class="cell" data-output-location="fragment">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_genotypes</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>1 multiallelic sites (0.072% out of 1390 total) detected and removed</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,389 × 7
     pos EUR_1_chr1 EUR_1_chr2 EUR_2_chr1 EUR_2_chr2 AFR_1_chr1 AFR_1_chr2
   &lt;int&gt;      &lt;int&gt;      &lt;int&gt;      &lt;int&gt;      &lt;int&gt;      &lt;int&gt;      &lt;int&gt;
 1  1001          0          1          0          0          1          0
 2  1389          0          1          0          0          1          0
 3  2425          0          1          0          0          1          0
 4  4345          1          0          1          1          0          1
 5  5130          0          0          0          0          1          0
 6  7057          1          0          1          1          1          1
 7  7962          1          1          1          1          1          1
 8  9380          1          1          1          1          1          1
 9  9752          0          0          0          0          0          1
10 11235          1          0          1          1          1          1
# … with 1,379 more rows</code></pre>
</div>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>